linode/manager
# feat: [M3-7673] - Add Account Limit support ticket to Linode Create flow

## Description 📝
This PR adds a new type of support ticket form: the account limit ticket. Currently, when creating a Linode and a user has:
- reached a limit on their account that prevents them from deploying the selected Linode plan
or
- reached the limit for the number of active services on their account

a conditional error notice is displayed. Clicking the contact support link the user is redirected to the Support page and now, in this PR, a new support ticket dialog with custom account-limit-related fields is displayed. Once submitted by the user, the support team can determine if the user is eligible to have access to more Linodes and/or larger plans. According to Customer Support, this is one of the top requests they receive from users. Similar to the [SMTP ticket](https://github.com/linode/manager/pull/8636) form, this change aims to improve the user experience and reduce the amount of back-and-forth between customers and Support when handling requests.
 
> [!NOTE]
> This PR *only* links this account limit support ticket in the Linode Create flow, to start. It will be added to the rest of the flows that users can encounter this notice in follow up PRs.

Follow up tasks:
- Be able to pass data from the submitted form to pre-populate a field like the Linode Plan selection (M3-8336)
- Add the Account Limit support ticket to other product create pages (M3-8335)

## Changes  🔄
- Adds a new support ticket type: `accountLimit`
- Adds new account limit support ticket fields
- Adds a validation schema
- Adds an integration test case
- Handles the already-globally-intercepted support ticket API error with JSX on the Linode Create flows to be able to link to this new ticket type when the expected error text is present in the error message

## Target release date 🗓️
7/22

## Preview 📷

| Description | Before  | After   | 
| ------- | ------- | ------- |
| Error notice | v1: ![Screenshot 2024-07-08 at 1 30 29 PM](https://github.com/linode/manager/assets/114685994/d3ff59de-45c1-4f7a-aafd-9c5cd56baa5c) ![Screenshot 2024-07-08 at 1 50 22 PM](https://github.com/linode/manager/assets/114685994/ab6d7b2b-fd9e-42ec-8781-09e04259ed5c) v2: ![Screenshot 2024-07-08 at 1 31 28 PM](https://github.com/linode/manager/assets/114685994/fd2317e2-7a9c-421e-8fae-fea9650e6fb8) ![Screenshot 2024-07-08 at 1 49 44 PM](https://github.com/linode/manager/assets/114685994/ff60744b-4769-4d84-880e-abc476833431)| No change |
| Linked support ticket | a general support ticket | ![Screenshot 2024-07-08 at 1 30 43 PM](https://github.com/linode/manager/assets/114685994/c5673329-ba0f-4772-bb1a-50513eeea19c) |
| Submitted ticket | N/A | ![Screenshot 2024-07-08 at 1 39 33 PM](https://github.com/linode/manager/assets/114685994/5c892170-2d25-4239-9ba8-88324dddf3d8) |

## How to test 🧪

### Prerequisites
(How to setup test environment)
- **Please test this on your dev account so we don't bother Customer Support with real tickets.**
- **Please test these changes on Linode Create flow v1 and v2.**
- Check out this PR and go to http://localhost:3000/linodes/create.
- Log into admin and be ready to change some things on your dev account.

### Verification steps
(How to verify changes)
- In admin, set your reputation to 0.
- Fill out the linode form and try to create the linode. 
- Confirm that the UI displays an error notice when user has reached a limit on their account that prevents them from deploying the selected Linode plan: `A limit on your account is preventing the deployment of the selected Linode plan. To request access to the plan, please contact Support and provide the Linode plan name.`

- In admin, reset your reputation to its previous value and set your thing limit to 0.
- Fill out the linode form and try to create the linode. 
- Confirm that the UI displays an error notice when user has reached a limit for the number of active services on their account: `You’ve reached a limit for the number of active services on your account. Please contact Support to request an increase and provide the total number of services you may need.`

For each of the above possible limit errors:
- Confirm that clicking "contact Support" redirects you to the support ticket landing page with a new Account Limit ticket open.
- Confirm that the ticket title is always "Account Limit Increase" and the First/Last Name and Company fields are pre-populated, if you have that data on your account. 
- Try to submit the ticket and confirm validation errors show for any required fields that aren't filled out.
- Fill out fields and submit the ticket. Confirm that you are redirected to the ticket details page, and the ticket body lists the questions and answers from the ticket.
- Confirm no regressions to other ticket types

- Remember to restore all values in admin to their state prior to testing!

- Verify that all test cases in Cypress pass and the added test coverage for this PR looks good:
```
yarn cy:run -s "cypress/e2e/core/helpAndSupport/open-support-ticket.spec.ts"
```

## As an Author I have considered 🤔

*Check all that apply*
